app/src/main/java/de/rochefort/childmonitor/DiscoverActivity.kt

```diff
@@ -0,0 +1,247 @@
+    private var discoveryListener: DiscoveryListener? = null
+    override fun onCreate(savedInstanceState: Bundle?) {
+        Log.i(TAG, "ChildMonitor start")
+        nsdManager = this.getSystemService(NSD_SERVICE) as NsdManager
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_discover)
+        val discoverChildButton = findViewById<Button>(R.id.discoverChildButton)
+        discoverChildButton.setOnClickListener { v: View? -> loadDiscoveryViaMdns() }
+        val enterChildAddressButton = findViewById<Button>(R.id.enterChildAddressButton)
+        enterChildAddressButton.setOnClickListener { v: View? -> loadDiscoveryViaAddress() }
+    }
+
+    private fun loadDiscoveryViaMdns() {
+        setContentView(R.layout.activity_discover_mdns)
+        startServiceDiscovery("_childmonitor._tcp.")
+    }
+
+    private fun loadDiscoveryViaAddress() {
+        setContentView(R.layout.activity_discover_address)
+        val connectButton = findViewById<Button>(R.id.connectViaAddressButton)
+        val addressField = findViewById<EditText>(R.id.ipAddressField)
+        val portField = findViewById<EditText>(R.id.portField)
+        val preferredAddress = getPreferences(MODE_PRIVATE).getString(PREF_KEY_CHILD_DEVICE_ADDRESS, null)
+        if (!preferredAddress.isNullOrEmpty()) {
+            addressField.setText(preferredAddress)
+        }
+        val preferredPort = getPreferences(MODE_PRIVATE).getInt(PREF_KEY_CHILD_DEVICE_PORT, -1)
+        if (preferredPort > 0) {
+            portField.setText(preferredPort.toString())
+        } else {
+            portField.setText("10000")
+        }
+        connectButton.setOnClickListener { v: View? ->
+            Log.i(TAG, "Connecting to child device via address")
+            val addressString = addressField.text.toString()
+            val portString = portField.text.toString()
+            if (addressString.isEmpty()) {
+                Toast.makeText(this@DiscoverActivity, R.string.invalidAddress, Toast.LENGTH_LONG).show()
+                return@setOnClickListener
+            }
+            val port: Int = try {
+                portString.toInt()
+            } catch (e: NumberFormatException) {
+                Toast.makeText(this@DiscoverActivity, R.string.invalidPort, Toast.LENGTH_LONG).show()
+                return@setOnClickListener
+            }
+            val preferencesEditor = getPreferences(MODE_PRIVATE).edit()
+            preferencesEditor.putString(PREF_KEY_CHILD_DEVICE_ADDRESS, addressString)
+            preferencesEditor.putInt(PREF_KEY_CHILD_DEVICE_PORT, port)
+            preferencesEditor.apply()
+            connectToChild(addressString, port, addressString)
+        }
+    }
+
+    override fun onDestroy() {
+        Log.i(TAG, "ChildMonitoring stop")
+        if (discoveryListener != null) {
+            Log.i(TAG, "Unregistering monitoring service")
+            nsdManager!!.stopServiceDiscovery(discoveryListener)
```

Comment:
using !! here breaks the null safety that kotlin provides. We should avoid this operator like the plague.
Rather use the ? operator, possibly in combination with let{}
On a side note: Many things that we just changed to local vars for null checking can now be done more elegantly using ?

Reply:
I suspect the reason it is done that way, that in fact it can never be null, because of the activity life-cycle.

`nsdManager` is set to a non-null value after `onCreate`. Probably the clean way to solve that is to extract that code to some class that is created in `onCreate`.

Reply:
From what I read, the standard practice is apparently to mark it as a `lateinit` variable.

Reply:
Yep, agreed, this makes sense here.
I am generally reluctant to use lateinit because in imho it is just another way to basically ignore the nullability of a field that make everything crash if the programmer's assumptions are wrong, but I agree with your lifecycle rationale.